### PR TITLE
un-revert "Simplify some views of Adjoint matrices"

### PR DESCRIPTION
**Pull request by mcabbott**
This copies https://github.com/JuliaLang/julia/pull/43719 to the new LinearAlgebra.jl repository (partly to see if this all works!) The original PR was https://github.com/JuliaLang/julia/pull/39467

Tests passed on CI in Feb, but I have not checked anything locally. 

(Closes https://github.com/JuliaLang/julia/pull/43719)

### CODE REVIEW

**Comment by KristofferC** [comment]
The windows CI error

```
ERROR: Unable to open agent private key path 'C:\secrets/agent.key'!  Make sure your agent has this file deployed within it!
```

happens spuriously and even happens on Julia repo tests sometimes.

**Comment by mcabbott** [comment]
Bump? After rebasing this passes CI.

The original PR was approved, but reverted because it broke the tests of some later PR (which was merged first). That code has since been edited not to rely on the absence of this optimisation, which is why tests pass again now.

(In fact I have a merge button, but am not sure what the rules are on my using it, and thus would prefer someone else to approve.)

File: src/adjtrans.jl [hunk 1]
```diff
@@ -382,6 +382,20 @@ parent(A::AdjOrTrans) = A.parent
 vec(v::TransposeAbsVec{<:Number}) = parent(v)
 vec(v::AdjointAbsVec{<:Real}) = parent(v)
 
+Base.reshape(v::TransposeAbsVec{<:Number}, ::Val{1}) = parent(v)
+Base.reshape(v::AdjointAbsVec{<:Real}, ::Val{1}) = parent(v)
+
+ # these make eachrow(A') produce simpler views
+@inline Base.unsafe_view(A::Transpose{<:Number, <:AbstractMatrix}, i::Integer, j::AbstractArray) =
+    Base.unsafe_view(parent(A), j, i)
```
Comment (jishnub) [inline src/adjtrans.jl:390]: These should probably call `view` on the parent, as the parent might have a specialized `view` implementation that avoids the `SubArray` wrapper.

File: src/adjtrans.jl [hunk 2]
```diff
@@ -382,6 +382,20 @@ parent(A::AdjOrTrans) = A.parent
 vec(v::TransposeAbsVec{<:Number}) = parent(v)
 vec(v::AdjointAbsVec{<:Real}) = parent(v)
 
+Base.reshape(v::TransposeAbsVec{<:Number}, ::Val{1}) = parent(v)
+Base.reshape(v::AdjointAbsVec{<:Real}, ::Val{1}) = parent(v)
+
+ # these make eachrow(A') produce simpler views
+@inline Base.unsafe_view(A::Transpose{<:Number, <:AbstractMatrix}, i::Integer, j::AbstractArray) =
```
Comment (jishnub) [inline src/adjtrans.jl:389]: Perhaps use the alias `TransposeAbsMat` like the `TransposeAbsVec` alias used in `reshape`?

File: src/adjtrans.jl [hunk 3]
```diff
@@ -382,6 +382,20 @@ parent(A::AdjOrTrans) = A.parent
 vec(v::TransposeAbsVec{<:Number}) = parent(v)
 vec(v::AdjointAbsVec{<:Real}) = parent(v)
 
+Base.reshape(v::TransposeAbsVec{<:Number}, ::Val{1}) = parent(v)
```
Comment (jishnub) [inline src/adjtrans.jl:385]: Should `reshape(v, ::Colon)` and such behave identically? Or is the `Val(1)` reshaping used in the view below? E.g.:
```julia
julia> v = 1:4
1:4

julia> reshape(v', :)
4-element reshape(adjoint(::UnitRange{Int64}), 4) with eltype Int64:
 1
 2
 3
 4

julia> reshape(v', Val(1))
1:4
```
It might be a good idea to have these all return a range, but that may be a separate PR.

File: src/adjtrans.jl (hunk 1 above)
Comment (jishnub) [reply src/adjtrans.jl:390]: Can't we also do something similar for transposes of `AbstractVector`s, with one `AbstractVector` index?

File: src/adjtrans.jl (hunk 3 above)
Comment (mcabbott) [reply src/adjtrans.jl:385]: Yea this PR is only about `view`. And some views call this (internal?) method with `Val`. 

The "supported" way to do that reshape is:
```julia
julia> vec(v')
1:4
```
The default definition is `vec(a::AbstractArray) = reshape(a,length(a))`, and at present `reshape(v', 4)` also has no shortcut.

File: src/adjtrans.jl (hunk 1 above)
Comment (mcabbott) [reply src/adjtrans.jl:390]: Probably. But that's a somewhat odd thing, and this PR doesn't try to cover all possible views -- just the ones which show up in `eachrow(A')`.

Comment (mcabbott) [reply src/adjtrans.jl:390]: Probably. But that's a somewhat odd thing, and this PR doesn't try to cover all possible views.

Comment (mcabbott) [reply src/adjtrans.jl:390]: I initially said the wrong thing, because I've forgotten how this all works, but in fact that case is already handled (and tested). It's why the `reshape(v, ::Val{1})` method is defined above.

I'm hesitant to start editing as I do not have the details fresh in my head, and do not wish to introduce fresh mistakes. The pervious PR reverted mine with no changes, this PR un-does the reversion.

File: src/adjtrans.jl (hunk 2 above)
Comment (mcabbott) [reply src/adjtrans.jl:389]: IDK why it's written this way, it's possible I followed nearby code which writes e.g. `Base.strides(A::Adjoint{<:Real, <:AbstractMatrix})` and does not seem to use TransposeAbsMat. Maybe I was concerned about order of definitions or something?

**Review by jishnub — APPROVED** [review]
Looks good to me
